### PR TITLE
sandbox: make sure qemu process not exit after stop

### PR DESCRIPTION
**Pull request by Ace-Tang**
We found qemu still living for a while after stopVM successful, this
will cause a problem when we use passthrough network. In stop
container/pod, network plugin try bind vf to host kernel driver, if qemu
still running, kernel got panic

bind vf to virtio-pci driver
```
0xffffffffa02f6d50 : __vfio_group_unset_container+0x0/0xc0 [vfio]
0xffffffffa02f6e39 : vfio_group_try_dissolve_container+0x29/0x30 [vfio]
0xffffffffa02f7192 : vfio_group_put_external_user+0x12/0x70 [vfio]
0xffffffffa056aff3 : kvm_vfio_group_put_external_user+0x23/0x40 [kvm]
0xffffffffa056b10b : kvm_vfio_destroy+0x3b/0x90 [kvm]
0xffffffffa0563421 : kvm_put_kvm+0x161/0x1e0 [kvm]
0xffffffffa0563511 : kvm_vm_release+0x21/0x30 [kvm]
```

kvm release container fd

```
0xffffffffa0045900 : virtnet_probe+0x0/0x0 [virtio_net]
0xffffffffa007d6e4 : virtio_dev_probe+0x144/0x1e0 [virtio]
0xffffffff814e2e37 : driver_probe_device+0x1c7/0x410 [kernel]
0xffffffff814e31fc : __device_attach_driver+0x8c/0x100 [kernel]
0xffffffff814e0957 : bus_for_each_drv+0x67/0xb0 [kernel]
0xffffffff814e2aed : __device_attach+0xdd/0x160 [kernel]
0xffffffff814e32b3 : device_initial_probe+0x13/0x20 [kernel]
0xffffffff814e1d22 : bus_probe_device+0x92/0xa0 [kernel]
0xffffffff814df889 : device_add+0x3c9/0x620 [kernel]
```

Fixes: #1945

Signed-off-by: Ace-Tang <aceapril@126.com>

### CODE REVIEW

**Comment by Ace-Tang** [comment]
Try close #1946. /cc @xs3c @gnawux @devimc 

After change wait time from 1s to 10s, the patch has run for half a month and work good.

File: virtcontainers/sandbox.go [hunk 1]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
 		s.Logger().Info("Not stopping VM")
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		// if use PhysicalEndpoint, wait for hypervisor exited
+		// https://work.aone.alibaba-inc.com/issue/20349397
```
Comment (devimc) [inline virtcontainers/sandbox.go]: is this an internal link? I can't access it

File: virtcontainers/sandbox.go [hunk 2]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
...
+			if running, _ := isShimRunning(pid); !running {
+				exited = true
+				break
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			networkLogger().Warnf("NOTE: hypervisor still running after stop %d", pid)
+			time.Sleep(200 * time.Millisecond)
+		}
+		if !exited {
+			networkLogger().Infof("NOTE: hypervisor %d not exited after 1s", pid)
```
Comment (devimc) [inline virtcontainers/sandbox.go]: 1s?  `50 * 200 milliseconds  = 10s` , right?

File: virtcontainers/sandbox.go [hunk 3]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
...
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		// if use PhysicalEndpoint, wait for hypervisor exited
+		// https://work.aone.alibaba-inc.com/issue/20349397
+		exited := false
+		for i := 0; i < 50; i++ {
+			if running, _ := isShimRunning(pid); !running {
```
Comment (devimc) [inline virtcontainers/sandbox.go]: here we are relying that the `shim` dies when the hypervisor dies, this is prone to bugs

File: virtcontainers/sandbox.go [hunk 4]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
 		s.Logger().Info("Not stopping VM")
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		// if use PhysicalEndpoint, wait for hypervisor exited
+		// https://work.aone.alibaba-inc.com/issue/20349397
+		exited := false
+		for i := 0; i < 50; i++ {
```
Comment (devimc) [inline virtcontainers/sandbox.go]: magic number

File: virtcontainers/sandbox.go [hunk 5]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
 		s.Logger().Info("Not stopping VM")
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		// if use PhysicalEndpoint, wait for hypervisor exited
```
Comment (devimc) [inline virtcontainers/sandbox.go]: missing condition?  `if PhysicalEndpoint {`

File: virtcontainers/sandbox.go [hunk 6]
```diff
@@ -1062,6 +1063,26 @@ func (s *Sandbox) stopVM() error {
...
+		// https://work.aone.alibaba-inc.com/issue/20349397
+		exited := false
+		for i := 0; i < 50; i++ {
+			if running, _ := isShimRunning(pid); !running {
+				exited = true
+				break
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			networkLogger().Warnf("NOTE: hypervisor still running after stop %d", pid)
```
Comment (devimc) [inline virtcontainers/sandbox.go]: Use `WithField`

File: virtcontainers/sandbox.go (hunk 2 above)
Comment (devimc) [inline virtcontainers/sandbox.go]: Use `WithField`

File: virtcontainers/sandbox.go (hunk 5 above)
Comment (Ace-Tang) [reply virtcontainers/sandbox.go]: Here I want add a explain, not condition.

**Comment by Ace-Tang** [comment]
Updated !

File: virtcontainers/sandbox.go [hunk 7]
```diff
@@ -1062,6 +1067,24 @@ func (s *Sandbox) stopVM() error {
 		s.Logger().Info("Not stopping VM")
 		return nil
 	}
+	pid := s.hypervisor.pid()
+
+	defer func() {
+		exited := false
```
Comment (devimc) [inline virtcontainers/sandbox.go]: I think this variable is not needed

File: virtcontainers/sandbox.go [hunk 8]
```diff
@@ -1062,6 +1067,24 @@ func (s *Sandbox) stopVM() error {
...
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				exited = true
```
Comment (devimc) [inline virtcontainers/sandbox.go]: just `return`

File: virtcontainers/sandbox.go [hunk 9]
```diff
@@ -1062,6 +1067,24 @@ func (s *Sandbox) stopVM() error {
...
+				break
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			networkLogger().WithField("pid", pid).Warn("NOTE: hypervisor still running after stopVM")
+			time.Sleep(waitVMExitTime * time.Millisecond)
+		}
+		if !exited {
```
Comment (devimc) [inline virtcontainers/sandbox.go]: and just log the error

File: virtcontainers/sandbox.go [hunk 10]
```diff
@@ -39,6 +40,10 @@ const (
 	// vmStartTimeout represents the time in seconds a sandbox can wait before
 	// to consider the VM starting operation failed.
 	vmStartTimeout = 10
+	// waitRetry represents retry times wait for vm exit
+	waitRetry = 50
+	// waitVMExit represents times wait for vm exit in one loop
+	waitVMExitTime = 200
```
Comment (devimc) [inline virtcontainers/sandbox.go]: since this is time, please convert it to milliseconds `200 * time.Millisecond` or mention that this value is in milliseconds

File: virtcontainers/sandbox.go [hunk 11]
```diff
@@ -1062,6 +1067,24 @@ func (s *Sandbox) stopVM() error {
...
+	defer func() {
+		exited := false
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				exited = true
+				break
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			networkLogger().WithField("pid", pid).Warn("NOTE: hypervisor still running after stopVM")
+			time.Sleep(waitVMExitTime * time.Millisecond)
```
Comment (devimc) [inline virtcontainers/sandbox.go]: `time.Sleep(waitVMExitTime)`

File: virtcontainers/sandbox.go [hunk 12]
```diff
@@ -1062,6 +1067,24 @@ func (s *Sandbox) stopVM() error {
...
+		}
+		if !exited {
+			networkLogger().WithField("pid", pid).Info("NOTE: hypervisor not exited after 10s")
```
Comment (devimc) [inline virtcontainers/sandbox.go]: `10s` can change depending of `waitRetry` and `waitVMExitTime`, could you calculate this value dynamically, something like `((waitRetry * waitVMExitTime) / time.Millisecond)`

File: virtcontainers/shim.go [hunk 13]
```diff
@@ -227,7 +227,7 @@ func startShim(args []string, params ShimParams) (int, error) {
 	return cmd.Process.Pid, nil
 }
 
-func isShimRunning(pid int) (bool, error) {
+func isProcRunning(pid int) (bool, error) {
```
Comment (devimc) [inline virtcontainers/shim.go]: Thanks, but this change should be part of a different commit

**Comment by Ace-Tang** [comment]
Hi, @devimc , I update. Should we skip the check if `s.hypervisor.stopSandbox()` get error ?

File: virtcontainers/sandbox.go [hunk 14]
```diff
@@ -1062,6 +1068,19 @@ func (s *Sandbox) stopVM() error {
...
+	defer func() {
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				return
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
+			time.Sleep(waitVMExitTime * time.Millisecond)
+		}
+		networkLogger().WithField("pid", pid).
+			Warnf("NOTE: hypervisor not exited after wait %ds", (waitRetry*waitVMExitTime)/(time.Second/time.Millisecond))
```
Comment (devimc) [inline virtcontainers/sandbox.go]: I think this calculation is wrong `(50 * 200)÷(1 / 1000000) = 10000000000` isn't it?

File: virtcontainers/sandbox.go [hunk 15]
```diff
@@ -1062,6 +1068,19 @@ func (s *Sandbox) stopVM() error {
...
+	}()
 
 	s.Logger().Info("Stopping VM")
 	return s.hypervisor.stopSandbox()
```
Comment (devimc) [inline virtcontainers/sandbox.go]: @Ace-Tang 
 >  I update. Should we skip the check if s.hypervisor.stopSandbox() get error ?

good catch!  I think we shouldn't skip that check, instead you can modify the signature of this function: 
```go
(s *Sandbox) stopVM() (err error) {
``` 
Log the error if `s.hypervisor.stopSandbox()` fails: 
```go
if err = s.hypervisor.stopSandbox(); err != nil {
   // Log error
}
```
 and finally modify the defer function

```go
defer func() {
    if err == nil {
        return
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
          .......
     // convert this is an error
     // networkLogger().WithField("pid", pid)...
    return fmt.Errorf("hypervisor not exited after ......")
```

what do you think ?

Comment (Ace-Tang) [reply virtcontainers/sandbox.go]: Even `s.hypervisor.stopSandbox()` return nil, the qemu may still living for a while, so I think `err
== nil`, the check is need, but if `s.hypervisor.stopSandbox()` return error, do we still need to check ? like

```
defer func() {
    if err != nil {
        // log error
        return 
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
....
     return fmt.Errorf("hypervisor not exited after ......")

```

Comment (devimc) [reply virtcontainers/sandbox.go]: @Ace-Tang 

> Even s.hypervisor.stopSandbox() return nil, the qemu may still living for a while, so I think err == nil, the check is need,

does `qmp` not shutdown the vm at all and return nil? 
if so, then `err == nil` condition is not needed, the error should be logged and the hypervisor killed, I think

```go
defer func() {
    if err != nil {
        // just log the error
    }
    for i := 0; i < waitRetry; i++ {
         if running, _ := isProcRunning(pid); !running {
               err = nil
               return
          }
.....
```

Comment (amshinde) [reply virtcontainers/sandbox.go]: @Ace-Tang Yes, if its the hypervisor taking some time to exit, without qmp returning any error, then we need to just log the error in case we get one and wait for the hypervisor to stop in both the error/non-error case.

Comment (Ace-Tang) [reply virtcontainers/sandbox.go]: > does qmp not shutdown the vm at all and return nil?

Yes, this is reason cause kernel panic, qemu still get vfio, and iommu try to unbind vf from vfio.
And Updated.

File: virtcontainers/sandbox.go [hunk 16]
```diff
@@ -1062,6 +1068,26 @@ func (s *Sandbox) stopVM() error {
...
+	defer func() {
+		if err != nil {
+			networkLogger().WithError(err).WithField("pid", pid).Warn("Fail to stop vm")
+		}
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				err = nil
+				return
+			}
+			syscall.Kill(pid, syscall.SIGKILL)
```
Comment (bergwolf) [inline virtcontainers/sandbox.go]: SIGKILL is not catchable. Please only send it once.

Comment (devimc) [reply virtcontainers/sandbox.go]: > Please only send it once.
 
@Ace-Tang what do you think? Have you seen cases where the first signal `SIGKILL` doesn't kill the hypervisor?

Comment (Ace-Tang) [reply virtcontainers/sandbox.go]: I am not seen qemu not be killed by SIGKILL

Comment (Ace-Tang) [reply virtcontainers/sandbox.go]: Updated, @devimc @bergwolf

File: virtcontainers/shim.go (hunk 13 above)
Comment (devimc) [reply virtcontainers/shim.go]: @Ace-Tang could you include this change in a different commit?

**Comment by Ace-Tang** [comment]
@devimc , I do split it to two commit, there are mistake here.  Sorry, :cry:  updated.

**Comment by devimc** [comment]
/test

@Ace-Tang thanks

**Comment by devimc** [comment]
@Ace-Tang please rebase this pr

**Comment by greg-bock** [comment]
Is this similar to #1946? The patch I linked to in this [comment](https://github.com/kata-containers/runtime/pull/1946#issuecomment-521018138) cleared up similar issues we had with sriov.

**Comment by devimc** [comment]
@greg-bock 

> Is this similar to #1946? 

both fix the same issue

> The patch I linked to in this comment cleared up similar issues we had with sriov.

Thanks for the patch, In your patch you are waiting for qemu to finish, but there are some scenarios where qemu hangs or takes too long to finish, maybe the `kill -9 $PID` approach is better for these cases, what do you think?

**Comment by greg-bock** [comment]
> In your patch you are waiting for qemu to finish, but there are some scenarios where qemu hangs or takes too long to finish, maybe the `kill -9 $PID` approach is better for these cases, what do you think?

It has been a while since we first patched this. I'm not sure if we ever saw qemu not eventually exit with our issue, but we've never seen it hang after we applied our patch.

The use of using `new_id` and `remove_id` driver override method in  `virtcontainers/device/drivers/vfio.go`is [non-deterministic](https://www.redhat.com/archives/libvir-list/2014-April/msg00222.html) and might also cause odd issues when using sriov. I [patched](https://gist.github.com/greg-bock/f4aa76058da5ba276efbac370e9e91f5) this to use the `driver_override` method which may have removed some other corner cases when using sriov. Note that this change is likely not safe, or at the very least could leave cruft around, for an in place upgrade unless you have another CNI plugin that can do the cleanup for the old `new_id` and `remove_id` style.

**Comment by devimc** [comment]
>  I patched this to use the driver_override

nice patch!, we can need it if we face odd issues with sriov

File: virtcontainers/sandbox.go [hunk 17]
```diff
@@ -1062,6 +1068,30 @@ func (s *Sandbox) stopVM() error {
...
+		for i := 0; i < waitRetry; i++ {
+			if running, _ := isProcRunning(pid); !running {
+				err = nil
+				return
+			}
+			time.Sleep(waitVMExitTime * time.Millisecond)
+		}
+		waitTime := (waitRetry * waitVMExitTime) / (time.Second / time.Millisecond)
+		networkLogger().WithField("pid", pid).
+			Warnf("NOTE: hypervisor not exited after wait %ds", waitTime)
```
Comment (jodh-intel) [inline virtcontainers/sandbox.go]: Nit: You could specify the wait time as a structured logger field:

```go
networkLogger().WithFields(logrus.Fields{"pid": pid, "wait-time": waitTime}).Warn("hypervisor still running after wait time")
```

**Comment by jodh-intel** [comment]
@Ace-Tang - All the CI's are failing this PR - could you tal?

**Comment by raravena80** [comment]
@Ace-Tang any updates on this PR? Thx!

**Comment by Ace-Tang** [comment]
closed， thanks for review
